django/db/models/fields/json.py

```diff
@@ -92,10 +96,17 @@ def from_db_value(self, value, expression, connection):
     def get_internal_type(self):
         return "JSONField"
 
-    def get_prep_value(self, value):
+    def get_db_prep_value(self, value, connection, prepared=False):
+        if hasattr(value, "as_sql"):
+            return value
+        if value == "null":
+            value = None
+        return connection.ops.adapt_json_value(value, self.encoder)
+
+    def get_db_prep_save(self, value, connection):
         if value is None:
             return value
-        return json.dumps(value, cls=self.encoder)
+        return self.get_db_prep_value(value, connection)
```

Comment:
While it would be another level of indirection, would it make sense to call `super()` here instead. The default impl calls `get_db_prep_value`.

Reply:
> The default impl calls get_db_prep_value.

It does but it also returns immediately when passed a compilable (`hasattr(value, "as_sql")`) so it can't be used as `JSONField.get_db_prep_value` needs to get access to `Value` objects to perform the `Value("null")` and friends deprecation properly.

https://github.com/django/django/blob/0e5532b3de24ce30dea621129ebdc9dc7d5b3fdf/django/db/models/fields/__init__.py#L922-L926